tools/prow-auto-bumper/main.go

```diff
@@ -0,0 +1,282 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// prow-auto-bumper finds stable Prow components version used by k8s,
+// and creates PRs updating them in knative/test-infra
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"math"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/knative/test-infra/shared/ghutil"
+)
+
+const (
+	org  = "kubernetes"
+	repo = "test-infra"
+	// PRHead is the head branch of k8s auto version bump PRs
+	// TODO(chaodaiG): using head branch querying is less ideal than using
+	// label `area/prow/bump`, which is not supported by Github API yet. Move
+	// to filter using this label once it's supported
+	PRHead = "k8s-ci-robot:autobump"
+	// PRBase is the base branch of k8s auto version bump PRs
+	PRBase = "master"
+	// Index for regex matching groups
+	imageImagePart = 1
+	imageTagPart   = 2
+	// Max difference away from target date
+	maxDelta = 2 * 24 // 2 days
+	// Safe duration is the smallest amount of hours a version stayed
+	safeDuration = 12 // 12 hours
+	maxRetry     = 3
+)
+
+var (
+	// matching            gcr.io /k8s-(prow|testimage)/(tide|kubekin-e2e|.*)    :vYYYYMMDD-HASH-EXTRA
+	imagePattern     = `\b(gcr\.io/k8s[a-z0-9-]{5,29}/[a-zA-Z0-9][a-zA-Z0-9_.-]+):(v[a-zA-Z0-9_.-]+)\b`
+	imageRegexp      = regexp.MustCompile(imagePattern)
+	imageLinePattern = fmt.Sprintf(`\s+[a-z]+:\s+"?'?%s"?'?`, imagePattern)
+	// matching   "-    image: gcr.io /k8s-(prow|testimage)/(tide|kubekin-e2e|.*)    :vYYYYMMDD-HASH-EXTRA"
+	imageMinusRegexp = regexp.MustCompile(fmt.Sprintf(`\-%s`, imageLinePattern))
+	// matching   "+    image: gcr.io /k8s-(prow|testimage)/(tide|kubekin-e2e|.*)    :vYYYYMMDD-HASH-EXTRA"
+	imagePlusRegexp = regexp.MustCompile(fmt.Sprintf(`\+%s`, imageLinePattern))
+	// Preferred time for candidate PR creation date
+	targetTime = time.Now().Add(-time.Hour * 7 * 24) // 7 days
+)
+
+// versions holds the version change for an image
+type versions struct {
+	oldVersion string
+	newVersion string
+	variant    string
+}
+
+// PRVersions contains PR and version changes in it
+type PRVersions struct {
+	imageVsMap map[string][]versions // map of image name: versions struct
+	dominantVs *versions             // Dominant versions in a PR
+	PR         *github.PullRequest
+}
+
+// Helper method for adding a newly discovered tag into pv
+func (pv *PRVersions) addTag(image, tag string, f func(*PRVersions, int)) {
+	if _, ok := pv.imageVsMap[image]; !ok {
+		pv.imageVsMap[image] = make([]versions, 0, 0)
+	}
+	_, variant := deconstructTag(tag)
+	iv := -1
+	for i, vs := range pv.imageVsMap[image] {
+		if vs.variant == variant {
+			iv = i
+		}
+	}
+	if -1 == iv {
+		pv.imageVsMap[image] = append(pv.imageVsMap[image], versions{variant: variant})
+		iv = len(pv.imageVsMap[image]) - 1
+	}
+	f(pv, iv)
```

Comment:
why not return iv here instead of taking a function and calling that? The caller can do:
```
i := addTag(image, tag)
pv.imageVsMap[image][i].newVersion = tag
```

Reply:
also, maybe rename addTag since it does not add tag anymore

Reply:
done